Sources/SwiftSyntax/SyntaxChildren.swift

```diff
@@ -10,305 +10,246 @@
 //
 //===----------------------------------------------------------------------===//
 
-/// Sequence over the layout children nodes of a raw node, that also provides
-/// its associated `AbsoluteSyntaxInfo`.
-struct RawSyntaxChildren: Sequence {
-  struct Iterator: IteratorProtocol {
-    let parent: RawSyntax
-    var nextChildInfo: AbsoluteSyntaxInfo
-
-    init(parent: AbsoluteRawSyntax) {
-      self.init(parent: parent.raw,
-        startFrom: parent.info.advancedToFirstChild())
-    }
-
-    init(parent: RawSyntax, startFrom: AbsoluteSyntaxInfo) {
-      self.parent = parent
-      self.nextChildInfo = startFrom
-    }
+public struct SyntaxChildrenIndex: Comparable {
+  fileprivate let offset: UInt32
+  fileprivate let indexInParent: UInt32
+  fileprivate let indexInTree: SyntaxIndexInTree
 
-    mutating func next() -> (RawSyntax?, AbsoluteSyntaxInfo)? {
-      let idx = Int(nextChildInfo.indexInParent)
-      guard idx < parent.numberOfChildren else {
-        return nil
-      }
-      let child = parent.child(at: idx)
-      let thisItem = (child, nextChildInfo)
-      nextChildInfo = nextChildInfo.advancedBySibling(child)
-      return thisItem
-    }
+  public static func <(lhs: SyntaxChildrenIndex, rhs: SyntaxChildrenIndex)
+      -> Bool {
+    return lhs.indexInParent < rhs.indexInParent
   }
 
-  private let node: AbsoluteRawSyntax
-
-  init(_ absoluteRaw: AbsoluteRawSyntax) {
-    self.node = absoluteRaw
+  fileprivate init(offset: UInt32, indexInParent: UInt32,
+                   indexInTree: SyntaxIndexInTree) {
+    self.offset = offset
+    self.indexInParent = indexInParent
+    self.indexInTree = indexInTree
   }
 
-  init(_ base: _SyntaxBase) {
-    self.init(base.data.absoluteRaw)
-  }
-
-  func makeIterator() -> Iterator {
-    return .init(parent: node)
+  init(_ absoluteSyntaxInfo: AbsoluteSyntaxInfo) {
+    self.offset = absoluteSyntaxInfo.offset
+    self.indexInParent = absoluteSyntaxInfo.indexInParent
+    self.indexInTree = absoluteSyntaxInfo.nodeId.indexInTree
   }
 }
 
-protocol AbsoluteRawSyntaxIteratorProtocol:
-  IteratorProtocol where Element == AbsoluteRawSyntax {
-}
+/// Collection that contains the child nodes of a raw node (including missing
+/// nodes), along with their associated `AbsoluteSyntaxInfo`.
+struct RawSyntaxChildren: BidirectionalCollection {
```

Comment:
We could also add a `distance(from:to:)` since we can do it in O(1) (difference of `indexInParent`)